util/fmtutil/format.go

```diff
@@ -0,0 +1,212 @@
+		labels[model.MetricNameLabel] = metricName + countStr
+		toTimeseries(wr, labels, timestamp, float64(m.GetSummary().GetSampleCount()))
+
+	case m.Histogram != nil:
+		metricName := labels[model.MetricNameLabel]
+		// Preserve metric name order with first bucket suffix timeseries then sum suffix timeserie and finally count suffix timeserie
+		// Add Histogram bucket timeseries
+		bucketLabels := make(map[string]string, len(labels)+1)
+		for key, value := range labels {
+			bucketLabels[key] = value
+		}
+		for _, b := range m.GetHistogram().Bucket {
+			bucketLabels[model.MetricNameLabel] = metricName + bucketStr
+			bucketLabels[model.BucketLabel] = fmt.Sprint(b.GetUpperBound())
+			toTimeseries(wr, bucketLabels, timestamp, float64(b.GetCumulativeCount()))
+		}
+		// Overwrite label model.MetricNameLabel for count and sum metrics
+		// Add Histogram sum timeserie
+		labels[model.MetricNameLabel] = metricName + sumStr
+		toTimeseries(wr, labels, timestamp, m.GetHistogram().GetSampleSum())
+		// Add Histogram count timeserie
+		labels[model.MetricNameLabel] = metricName + countStr
+		toTimeseries(wr, labels, timestamp, float64(m.GetHistogram().GetSampleCount()))
+
+	case m.Untyped != nil:
+		toTimeseries(wr, labels, timestamp, m.GetUntyped().GetValue())
+	default:
+		err = errors.New("unsupported metric type")
+	}
+	return err
+}
+
+func makeLabels(labelsMap map[string]string) []prompb.Label {
+	// build labels name list
+	sortedLabelNames := make([]string, 0, len(labelsMap))
+	for label := range labelsMap {
+		sortedLabelNames = append(sortedLabelNames, label)
+	}
+	// sort labels name in lexicographical order
+	sort.Strings(sortedLabelNames)
+
+	var labels []prompb.Label
+	for _, label := range sortedLabelNames {
+		labels = append(labels, prompb.Label{
+			Name:  label,
+			Value: labelsMap[label],
+		})
+	}
+	return labels
+}
+
+func makeLabelsMap(m *dto.Metric, metricName string, extraLabels map[string]string) map[string]string {
+	// build labels map
+	labels := make(map[string]string, len(m.Label)+len(extraLabels))
+	labels[model.MetricNameLabel] = metricName
+
+	// add extra labels
+	for key, value := range extraLabels {
+		labels[key] = value
+	}
+
+	// add metric labels
+	for _, label := range m.Label {
+		labelname := label.GetName()
+		if labelname == model.JobLabel {
+			labelname = fmt.Sprintf("%s%s", model.ExportedLabelPrefix, labelname)
+		}
+		labels[labelname] = label.GetValue()
+	}
+
+	return labels
+}
+
+// ParseMetricsTextReader consumes an io.Reader and returns the MetricFamily.
+func ParseMetricsTextReader(input io.Reader) (map[string]*dto.MetricFamily, error) {
+	var parser expfmt.TextParser
+	mf, err := parser.TextToMetricFamilies(input)
+	if err != nil {
+		return nil, err
+	}
+	return mf, nil
+}
+
+// ParseMetricsTextAndFormat return the data in the expected prometheus metrics write request format.
+func ParseMetricsTextAndFormat(input io.Reader, labels map[string]string) (*prompb.WriteRequest, error) {
```

Comment:
I think the ergonomics of this function are rather awkward. Since this is only ever called once I think it would be better to just have the caller convert the text format to dto.MetricsFamily (as in call `ParseMetricsTextReader`) and then create the request (call `CreateWriteRequest`, maybe a bit more descriptive name though).